### PR TITLE
[shopify-buy] Refactor shopify-buy types definitions

### PR DESCRIPTION
**Pull request by martin-badin**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test shopify-buy`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/Shopify/js-buy-sdk>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## shopify-buy ([<kbd>unpkg</kbd>](https://unpkg.com/browse/shopify-buy@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'shopify-buy'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 81d34e7b2170ccac2845bbb3b16b5ebebb509749
</p>

File: types/shopify-buy/tslint.json
Comment (gabritto) [inline types/shopify-buy/tslint.json:4]: Why do you need to turn off this rule?

Comment (martin-badin) [reply types/shopify-buy/tslint.json:4]: typescript expected `export =` the end of file. I am not sure, because the library uses `export default`. Do you have any idea how to remove this rule? I would like to remove that but I am not sure how.

Comment (gabritto) [reply types/shopify-buy/tslint.json:4]: I looked at the library, and what happens is that, in `package.json`, `main` points to `index.js`, and `index.js` uses `module.exports = ...`, therefore `export =` should be used in the .d.ts file. However, the same `package.json` also has a `jsnext:main` field pointing to `index.es.js`, which indeed uses `export default`. Now, since `jsnext:main` is a field in `package.json` that seems to be supported by specific bundlers, we don't have a way of supporting that. If the original package used export maps to provide both commonjs and esm entry points, then we could do the same here...
In light of that, I think you should use `export =` in the .d.ts file, but I'm open to hearing your opinion on that.

Comment (martin-badin) [reply types/shopify-buy/tslint.json:4]: I would like to export all types as well. Not just the one method. Is there any way how to do it?

Comment (gabritto) [reply types/shopify-buy/tslint.json:4]: That depends. How are the types going to be imported/used? The way the typings were written before is one of the ways, where you have something like:

```ts
declare namespace Foo {
    type Baz = { b: "b" };
}
declare function Foo(a: number, b: number): number;
export = Foo;
```

and then when you do `import Foo = require("foo")`, you can access types like so: `Foo.Baz`.
But maybe your question is about how to do re-exports of the types that are now being imported into the main `index.d.ts`?

Comment (martin-badin) [reply types/shopify-buy/tslint.json]: thanks for your review and help but I am still not able to export types. Please could you write an example of how should it be right?

Comment (gabritto) [reply types/shopify-buy/tslint.json]: I have tried thinking of ways of doing it, but I couldn't really think of anything besides manually importing the types from the other files and exporting them from inside the `ShopifyBuy` namespace in `index.d.ts`. Maybe it's just too complicated to refactor this package in the way you intended to? 😕

Comment (martin-badin) [reply types/shopify-buy/tslint.json]: @gabritto Thanks a lot for your help. Could you please review it?

**Review by andrewbranch — CHANGES_REQUESTED** [review]
I don’t really understand this change. None of the files you’ve added [actually exist](https://unpkg.com/browse/shopify-buy@2.17.1/) in the package. A module `.d.ts` file implies the existence of a `.js` file by the same name—on DefinitelyTyped, unfortunately we can’t just invent modules for the purposes of organizing types. (Concretely, this is because users will be allowed, or even prompted, to import from module specifiers like `"shopify-buy/checkout-collection"`, which does not exist at runtime and therefore will crash.)

**Comment by martin-badin** [comment]
> 

please see GitHub project https://github.com/Shopify/js-buy-sdk

**Comment by andrewbranch** [comment]
What about it?

**Comment by martin-badin** [comment]
You mentioned you don't understand this change. So this change follows the project structure, not the files on your mentioned page.

**Comment by andrewbranch** [comment]
The project structure doesn’t matter. An `@types` package that types an npm package has to reflect the structure of the npm package as it exists in the users’ node_modules (and I explained why in my previous message). Sorry, we can’t accept this change.

**Comment by martin-badin** [comment]
Sometimes matter on project structure. Sometimes node_modules folder contains the project structure. But yes. In this case, you are right. There are no files like this. I will fix it

**Comment by andrewbranch** [comment]
It seems like these are completely rewritten, not just “refactored.” Can any owners @openminder, @straiforos, @kilinkis, @chrisworman-pela, @MaciekBaron review? Why are there so many changes? Did the library fundamentally change in the course of a few minor version bumps? Were the old types wrong? If you want someone to spend the time to review 2k lines of changes, it might help to explain why it’s worth it. But it still seems likely that none of the current owners will be up for such a large review, in which case you’ll probably have better luck splitting something like this into smaller pieces.

**Comment by martin-badin** [comment]
Thanks, @andrewbranch. Short explanation:
- I think the previous version was very light (therefore almost 2k line)
- I used official documentation https://shopify.dev/api/storefront as a reference
- I used the same name of interfaces, etc as in the official documentation
- I marked deprecated properties based on official documentation

I would rather wait for a review from the owners because I can't imagine how to split this into smaller pieces.

**Comment by martin-badin** [comment]
Can any owners @openminder, @straiforos, @kilinkis, @chrisworman-pela, @MaciekBaron review?

**Review by gabritto — APPROVED** [review]

**Comment by martin-badin** [comment]
Ready to merge

**Pull request merged**
